### PR TITLE
NIFI-13877: JdbcCatalog connections are not closed properly in PutIceberg

### PR DESCRIPTION
**Pull request by mark-bathori**
# Summary

[NIFI-13877](https://issues.apache.org/jira/browse/NIFI-13877)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Review by exceptionfactory — COMMENTED** [review]
Thanks for addressing this issue @mark-bathori. Although this change looks straightforward, making use of `BaseMetastoreCatalog` as a `Closeable` in try-with-resources, what do you think about simply checking if the `Catalog` instance implements `Closeable` and handling that in a `finally` block in `loadTable`? That would minimize the scope of changes.

**Comment by mark-bathori** [comment]
Thanks for the review @exceptionfactory. Unfortunately the catalog can't be closed within the `loadTable` method because the jdbc connection is used in later part of the processor when the commit happens to the table (this is used via the table object). The finally block could be added to the lower try-catch block so the `Catalog` type could be kept but the `loadCatalog` extraction would be still needed to make the catalog object accessible.

Currently every supported `Catalog` implementation extended from `BaseMetastoreCatalog` but if there will be other catalog support added in the future that is not extended from that class e.g. RESTCatalog, then your suggested solution will be needed (checking if the `Catalog` instance implements `Closeable` and handling that in a finally block)
Let me know what you think.

**Comment by exceptionfactory** [comment]
> Thanks for the review @exceptionfactory. Unfortunately the catalog can't be closed within the `loadTable` method because the jdbc connection is used in later part of the processor when the commit happens to the table (this is used via the table object). The finally block could be added to the lower try-catch block so the `Catalog` type could be kept but the `loadCatalog` extraction would be still needed to make the catalog object accessible.
> 
> Currently every supported `Catalog` implementation extended from `BaseMetastoreCatalog` but if there will be other catalog support added in the future that is not extended from that class e.g. RESTCatalog, then your suggested solution will be needed (checking if the `Catalog` instance implements `Closeable` and handling that in a finally block) Let me know what you think.

Thanks for the reply @mark-bathori. With the growing popularity of REST Catalog implementations, particularly Apache Polaris, I expect this feature to be requested in the future. It may also require more significant refactoring, but I think keeping it in mind, in this particular change, is worth doing, so I recommend avoiding the switch to `BaseMetastoreCatalog`.

File: nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-common/src/main/java/org/apache/nifi/processors/iceberg/catalog/ClosableCatalog.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Wrapper class for Iceberg catalogs
+ */
+public class ClosableCatalog implements Catalog, Closeable {
```
Comment (exceptionfactory) [inline nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-common/src/main/java/org/apache/nifi/processors/iceberg/catalog/ClosableCatalog.java:36]:
```suggestion
public class CloseableCatalog implements Catalog, Closeable {
```

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the updates @mark-bathori, however, I think the `ClosableCatalog` wrapper is a step in a different direction. The wrapper interface will require additional maintenance as new methods may be added in future Iceberg versions. Did you consider defining `Catalog` outside of the `try` block and simply do an `instanceof` check in a new `finally` block?

**Comment by exceptionfactory** [comment]
Thanks for the latest update @mark-bathori, this approach looks good. Planning to merge pending successful CI builds.

**Review by exceptionfactory — APPROVED** [review]
+1 merging, thanks @mark-bathori!

**Pull request merged**